pkg/proxy/runner/bounded_frequency_runner_test.go

```diff
@@ -17,437 +17,474 @@ limitations under the License.
-func (ft *fakeTimer) advance(d time.Duration) {
-	ft.lock.Lock()
-	defer ft.lock.Unlock()
+	// Now do another run
+	runner.Run()
+	if !obj.waitForValue(4) {
+		t.Errorf("expected four calls, got %d", obj.calls())
+	}
 
-	ft.now = ft.now.Add(d)
-	if ft.active && !ft.now.Before(ft.timeout) {
-		ft.active = false
-		ft.c <- ft.timeout
+	// Retry was cancelled because we already ran
+	fakeClock.Step(4 * time.Second)
+	if !obj.waitForValue(4) {
+		t.Errorf("expected four calls, got %d", obj.calls())
 	}
-}
 
-// return the calling line number (for printing)
-// test the timer's state
-func checkTimer(name string, t *testing.T, upd timerUpdate, active bool, next time.Duration) {
-	if upd.active != active {
-		t.Fatalf("%s: expected timer active=%v", name, active)
+	runner.Run()
+	if !obj.waitForValue(5) {
+		t.Errorf("expected five calls, got %d", obj.calls())
 	}
-	if active && upd.next != next {
-		t.Fatalf("%s: expected timer to be %v, got %v", name, next, upd.next)
+	runner.Retry()
+	if !obj.waitForValue(5) {
+		t.Errorf("expected five calls, got %d", obj.calls())
 	}
-}
-
-// test and reset the receiver's state
-func checkReceiver(name string, t *testing.T, receiver *receiver, expected bool) {
-	triggered := receiver.reset()
-	if expected && !triggered {
-		t.Fatalf("%s: function should have been called", name)
-	} else if !expected && triggered {
-		t.Fatalf("%s: function should not have been called", name)
+	// Call Run again before minInterval passes
+	fakeClock.Step(100 * time.Millisecond) // rel=100ms
+	runner.Run()
+	if !obj.waitForValue(5) {
+		t.Errorf("expected five calls, got %d", obj.calls())
+	}
+	// Deferred run will run after minInterval passes
+	fakeClock.Step(900 * time.Millisecond) // rel=1000ms
+	if !obj.waitForValue(6) {
+		t.Errorf("expected six calls, got %d", obj.calls())
+	}
+	// Retry was cancelled because we already ran
+	fakeClock.Step(4 * time.Second) // rel=4s since run, 5s since RetryAfter
+	if !obj.waitForValue(6) {
+		t.Errorf("expected six calls, got %d", obj.calls())
+	}
+	// Rerun happens after maxInterval
+	fakeClock.Step(5 * time.Second) // rel=9s since run, 10s since RetryAfter
+	if !obj.waitForValue(6) {
+		t.Errorf("expected six calls, got %d", obj.calls())
+	}
+	fakeClock.Step(time.Second) // rel=10s since run
+	if !obj.waitForValue(7) {
+		t.Errorf("expected seven calls, got %d", obj.calls())
 	}
 }
 
-// Durations embedded in test cases depend on these.
-var minInterval = 1 * time.Second
-var maxInterval = 10 * time.Second
-
-func waitForReset(name string, t *testing.T, timer *fakeTimer, obj *receiver, expectCall bool, expectNext time.Duration) {
-	upd := <-timer.updated // wait for stop
-	checkReceiver(name, t, obj, expectCall)
-	checkReceiver(name, t, obj, false) // prove post-condition
-	checkTimer(name, t, upd, false, 0)
-	upd = <-timer.updated // wait for reset
-	checkTimer(name, t, upd, true, expectNext)
-}
+func Test_BoundedFrequencyRunnerRetry_RealClock(t *testing.T) {
```

Comment:
Why? The whole point of using `clock.Clock` is so we can avoid needing to do racy tests with the real clock.

Reply:
because fakeclock are not reals and are not  a one to one replacement of a real clock, I would not test something only using fake clocks, that will not give you real coverage

Reply:
how are they not real? The _entire point_ of FakeClock is to be usable as a replacement for the real clock in unit tests

Reply:
well, I can assure you they are not, is fake, so it is as close and good as the people implemented it ... with all the possible bugs and limitations, check the implementation if you are interested ... real clocks are much more complex, I went with clayton deep on this https://github.com/kubernetes/kubernetes/pull/107826  ... fake clock is a mock, and as a mock it works well, but is not a replacement ...